test/micro/org/openjdk/bench/java/math/Shared.java

```diff
@@ -0,0 +1,134 @@
+
+    // General note
+    // ============
+    //
+    // Isn't there a simple way to get a BigInteger of the specified number
+    // of bits of magnitude? It does not seem like it.
+    //
+    // We cannot create a BigInteger of the specified number of bytes,
+    // directly and *cheaply*. This constructor does not do what you
+    // might think it does:
+    //
+    //      BigInteger(int numBits, Random rnd)
+    //
+    //  The only real direct option we have is this constructor:
+    //
+    //      BigInteger(int bitLength, int certainty, Random rnd)
+    //
+    //  But even with certainty == 0, it is not cheap. So, create the
+    //  number with the closest number of bytes and then shift right
+    //  the excess bits.
+
+    private Shared() {
+        throw new AssertionError("This is a utility class");
+    }
+
+    //
+    // Creates a pair of same sign numbers x and y that minimally differ in
+    // magnitude.
+    //
+    // More formally: x.bitLength() == nBits and x.signum() == y.signum()
+    // and either
+    //
+    //   * y.bitLength() == nBits, and
+    //   * x.testBit(0) != y.testBit(0)
+    //
+    // or
+    //
+    //   * y.bitLength() == nBits + 1
+    //
+    // By construction, such numbers are unequal to each other, but the
+    // difference in magnitude is minimal. That way, the comparison
+    // methods, such as equals and compareTo, are forced to examine
+    // the _complete_ number representation.
+    //
+    // Assumptions on BigInteger mechanics
+    // ===================================
+    //
+    // 1. bigLength() is not consulted with for short-circuiting; if it is,
+    //    then we have a problem with nBits={0,1}
+    // 2. new BigInteger(0, new byte[]{0}) and new BigInteger(1, new byte[]{1})
+    //    are not canonicalized to BigInteger.ZERO and BigInteger.ONE,
+    //    respectively; if they are, then internal optimizations might be
+    //    possible (BigInteger is not exactly a value-based class).
+    // 3. Comparison and equality are checked from the most significant bit
+    //    to the least significant bit, not the other way around (for
+    //    comparison it seems natural, but not for equality). If any
+    //    of those are checked in the opposite direction, then the check
+    //    might short-circuit.
+    //
+    public static Pair createPair(int nBits) {
+        if (nBits < 0) {
+            throw new IllegalArgumentException(String.valueOf(nBits));
+        } else if (nBits == 0) {
+            var zero = new BigInteger(nBits, new byte[0]);
+            var one = new BigInteger(/* positive */ 1, new byte[]{1});
+            return new Pair(zero, one);
+        } else if (nBits == 1) {
+            var one = new BigInteger(/* positive */ 1, new byte[]{1});
+            var two = new BigInteger(/* positive */ 1, new byte[]{2});
+            return new Pair(one, two);
+        }
+        int nBytes = (nBits + 7) / 8;
+        var r = new Random();
```

Comment:
Some have a preference for providing a seed for `Random` instances in micros. Either hard-coded or through a `@Param` (I find this a bit excessive). Doing so might reduce run-to-run noise.

Reply:
AFAIK, if you need reproducible randoms in tests, you should add the tags:
```
 * @key randomness
 * @library /test/lib
```
and initialize your random generator with
```
import jdk.test.lib.RandomFactory;
...
Random rnd = RandomFactory.getRandom();
```
This prints the seed to STDOUT.

Reply:
> AFAIK, if you need reproducible randoms in tests, you should add the tags:
> 
> ```
>  * @key randomness
>  * @library /test/lib
> ```
> 
> and initialize your random generator with
> 
> ```
> import jdk.test.lib.RandomFactory;
> ...
> Random rnd = RandomFactory.getRandom();
> ```
> 
> This prints the seed to STDOUT.

It'd be true for tests, but not for benchmarks. Also: JTREG test tags for benchmarks?

Reply:
My bad. Forget my note

Reply:
> Some have a preference for providing a seed for `Random` instances in micros. Either hard-coded or through a `@Param` (I find this a bit excessive). Doing so might reduce run-to-run noise.

I hear you, but unless your opinion is strong, I'll leave it as is.

Reply:
Hard-coding a seed would remove the run-to-run randomness, too, so would remove the need the `randomness` tag if this utility were to be used from a functional test. But yes, the reason to not do true randomness when setting up data for micros is to ensure we don't get multimodal results from arbitrary differences in the data setup. There are other ways around that, e.g. randomize data continually between iterations, run more forks to get more samples about flicker etc. A lot of work with an end result which is unlikely to add little more quality than a "random" set of data that we keep static from run-to-run.